File: java/BinaryTreeInorderTraversal.java
/*
 * Filename: BinaryTreeInorderTraversal.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 29 May 2015 09:30:01 AM EDT
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public List<Integer> inorderTraversal(TreeNode root) {
		List<Integer> res = new ArrayList<Integer>();

		if (roor == null) {
			return res;
		}

		inorderTraversalAux(root, res);
		return res;
	}

	private void inorderTraversalAux(TreeNode root, List<Integer> res) {
		if (root == null) {
			inorderTraversalAux(root.left, res);
			res.add(root.val);
			inorderTraversalAux(root.left, res);
		}

		return;
	}
}

File: cpp/binary_tree_inorder_traversal.cpp
/*
 * Filename: binary_tree_inorder_traversal.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Thu 12 Nov 2015 11:23:19 PM EST
 * Description: 
 */

#include <iostream>
#include <stack>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution
{
	public:
		vector<int> inorderTraversal(TreeNode* root)
		{
			vector<int> res;
			stack<TreeNode*> stack;

			while (!stack.empty() || root != NULL)
			{ 
				if (root != NULL)
				{
					stack.push(root);
					root = root->left;
				}
				else
				{
					root = stack.top();
					stack.pop();
					res.push_back(root->val);
					root = root->right;
				}
			}

			return res;
		}
};
